releng/tools/rmake

```diff
@@ -58,86 +80,142 @@ def printTargets():
-        suitesToClean.update(aSuites)
+        suitesToClean.update(allowedSuites)
         used = True
     if pattern.startswith("clean-"):
         suitesToClean.add(pattern[6:])
         used = True
     if not used:
         sys.exit(f"Invalid target pattern specified: '{pattern}'")
 
-targetsToCompile = list(targetsToCompile)
-targetsToCompile.sort(key=lambda x: x.suite)
+def build(patterns):
+    if not allAreTargets(patterns):
+        sys.exit("Invalid usage!\n" + usage)
 
-for s in suitesToClean:
-    cleanSuite(s)
+    aSuites = configuredSuites()
+    aTargets = availableTargets()
+
+    suitesToCompile = set()
+    targetsToCompile = set()
+    suitesToClean = set()
+
+    for pattern in patterns:
+        matchPattern(pattern, aSuites, aTargets, suitesToCompile,
+            targetsToCompile, suitesToClean)
+
+    targetsToCompile = list(targetsToCompile)
+    targetsToCompile.sort(key=lambda x: x.suite)
+
+    for s in suitesToClean:
+        cleanSuite(s)
+
+    for s in suitesToCompile:
+        compileSuite(s)
+
+    compileTargets(targetsToCompile)
+
+def buildAll():
+    for suite in configuredSuites():
+        if os.path.exists(os.path.join(buildDir, suite)):
+            compileSuite(suite)
+    sys.exit(0)
+
+def test(patterns):
+    aSuites = configuredSuites()
+    aTargets = availableTestTargets()
+
+    if patterns[0] == "--all":
+        patterns = [x.target for x in aTargets]
+    if not allAreTargets(patterns):
+        sys.exit("XInvalid usage!\n" + usage)
+
+    suitesToCompile = set()
+    targetsToCompile = set()
+    suitesToClean = set()
+    for pattern in patterns:
+        matchPattern(pattern, aSuites, aTargets, suitesToCompile,
+            targetsToCompile, suitesToClean)
+
+    for s in suitesToCompile:
+        targetsToCompile.union(expandSuite(s, aTargets))
+
+    targetsToCompile = list(targetsToCompile)
+    targetsToCompile.sort(key=lambda x: x.suite)
+
+    for s in suitesToClean:
+        cleanSuite(s)
+
+    compileTargets(targetsToCompile)
+
+    for t in targetsToCompile:
+        runTest(t)
+
+def main():
+    if len(sys.argv) < 2:
+        sys.exit("Invalid usage!\n\n" + usage)
+
+    command = sys.argv[1]
+
+    if command == "--list":
+        if len(sys.argv) != 2:
+            sys.exit("Invalid usage!\n\n" + usage)
+        printTargets()
+        sys.exit(0)
+
+    if command == "--doc":
+        if len(sys.argv) != 2:
+            sys.exit("Invalid usage!\n\n" + usage)
+        try:
+            tool = os.path.join(root, "releng", "tools", "_build_doc.sh")
+            subprocess.check_call([tool])
+            sys.exit(0)
+        except subprocess.SubprocessError as e:
+            sys.exit(1)
+        except KeyboardInterrupt:
+            sys.exit(0)
+
+    if command == "--help":
+        if len(sys.argv) != 2:
+            sys.exit("Invalid usage!\n" + usage)
+        print(usage)
+        sys.exit(0)
+
+    if command == "--test":
+        test(sys.argv[2:])
```

Comment:
You allow `rmake --test --all nonsense` - which is different then for other flags.